### PR TITLE
Initial File format update changes + basic unit tests.

### PR DESCRIPTION
**Pull request by bogawa-r7**
## Description

Initial change for file format change. By default this still writes v1 records rather than v2 records, although that can be switched (and there is an API for v2 record writing). Reader checks whether it is reading a V1 or V2 record.

## Testing

Simple unit test only, not complete.

### CODE REVIEW

File: armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java [hunk 1]
```diff
@@ -307,12 +309,354 @@ private void load(DataInputStream inputStream) throws IOException {
...
+    sections.add(metadataSection);
+
+    // Send entity dictionary
+    List<Path> tempPaths = new ArrayList<>();
+    //totalBytes += 8;
+    boolean success = false;
+    try {
+      sections.add(computeEntityDictionarySection(compress, tempPaths));
+      sections.add(computeValueDictionarySection(compress, tempPaths));
+      sections.add(computeEntityIndexSection(compress, tempPaths));
```
Comment (alee-r7) [inline armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java:502]: Lets write the RowGroup first for v2 then the entity index. Lets see what happens.

**Review by alee-r7 — COMMENTED** [review]
I think we should force all writers to not write to v1 anymore. Also lets change the order of RG and entity index. I originally wanted to do this, so it'll test the format change and allow for the optimization I'm thinking of.

File: armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java [hunk 2]
```diff
@@ -307,12 +309,354 @@ private void load(DataInputStream inputStream) throws IOException {
...
+  }
+
+  private Section computeRowGroupSection(Compression compress, List<Path> tempPaths)
+     throws IOException
+  {
+    return computeSectionCompressible(compress, tempPaths, "rowgroup-temp_",
+       ColumnFileSection.ROWGROUP, rowGroupWriter, null);
+  }
+
+  private void compressToTempFile2(List<Path> tempPaths, SectionBlobPair pair, String tempFileName, Component component)
```
Comment (alee-r7) [inline armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java:538]: Prefer a  better name.

File: armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java [hunk 3]
```diff
@@ -307,12 +309,354 @@ private void load(DataInputStream inputStream) throws IOException {
...
+      return new SingleSection(this.sectionType, this.outputStream.toByteArray());
+    }
+  }
+
+  static interface SectionBlob {
+    int getLength();
+    InputStream getInputStream();
+  }
+
+  static class SingleSectionBlob implements SectionBlob {
```
Comment (alee-r7) [inline armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java:352]: I think we could probably rename this to like ByteArraySectionBlob

File: armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java [hunk 4]
```diff
@@ -307,12 +309,354 @@ private void load(DataInputStream inputStream) throws IOException {
...
+      return byteArray.length;
+    }
+
+    @Override public InputStream getInputStream()
+    {
+      return new ByteArrayInputStream(this.byteArray);
+    }
+  }
+
+  static class InputStreamLengthSectionBlob implements SectionBlob {
```
Comment (alee-r7) [inline armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java:371]: I think you can remove the length part of this name and call it InputStreamSectionBlob

File: armor-base/src/main/java/com/rapid7/armor/columnfile/ColumnFileSection.java [hunk 5]
```diff
@@ -1,10 +1,30 @@
...
+
 public enum ColumnFileSection {
-  HEADER,
-  METADATA,
-  ENTITY_DICTIONARY,
-  VALUE_DICTIONARY,
-  ENTITY_INDEX,
-  ROWGROUP
+  HEADER(-1),
+  TOC(-1), // only in V2
```
[the remark this answers is not in the document]
Comment (bogawa-r7) [reply armor-base/src/main/java/com/rapid7/armor/columnfile/ColumnFileSection.java:7]: These sections can't appear in the table of contents, so they don't need to have an ID for the table of contents.

File: armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java [hunk 6]
```diff
@@ -307,12 +309,354 @@ private void load(DataInputStream inputStream) throws IOException {
...
-  
+
   public StreamProduct buildInputStream(Compression compress) throws IOException {
+    return buildInputStream(compress, DEFAULT_VERSION);
+  }
+
+  public StreamProduct buildInputStream(Compression compress, Constants.ColumnFileFormatVersion version) throws IOException {
+    switch (version) {
+      case VERSION_1:
+        return buildInputStreamV1(compress);
```
[the remark this answers is not in the document]
Comment (bogawa-r7) [reply armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java]: removed the writing of V1.

File: armor-base/src/main/java/com/rapid7/armor/columnfile/ColumnFileReader.java [hunk 7]
```diff
@@ -31,57 +34,111 @@ public void read(DataInputStream dataInputStream, ColumnFileListener listener) t
...
+      }
+      totalBytesRead += readBytes;
     }
-    int rgCompressed = dataInputStream.readInt();
-    int rgOriginal = dataInputStream.readInt();
-    if (listener != null) {
-      readBytes = listener.columnFileSection(ColumnFileSection.ROWGROUP, metadata, dataInputStream, rgCompressed, rgOriginal);
+  }
+
+  private int readSection(DataInputStream dataInputStream, ColumnFileListener listener, ColumnFileSection valueDictionary)
```
[the remark this answers is not in the document]
Comment (bogawa-r7) [reply armor-base/src/main/java/com/rapid7/armor/columnfile/ColumnFileReader.java:106]: fixed

File: armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java (hunk 1 above)
Comment (bogawa-r7) [reply armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java:516]: per our discussion, the code assumes that EntityIndex exists prior to reading RowGroup

File: armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java (hunk 2 above)
Comment (bogawa-r7) [reply armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java]: now is compressToTempFile

File: armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java (hunk 3 above)
Comment (bogawa-r7) [reply armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java]: fixed.

File: armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java (hunk 4 above)
Comment (bogawa-r7) [reply armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java]: Renamed.

**Review by alee-r7 — APPROVED** [review]
Lets add the dummy test and go ahead and merge. Let me know when you merge it in, I'll start the process of releasing.

**Pull request merged**